File: src/py/simulator.py
import market_simulator
from helper.data_types import Orders, MarketData
from helper.place_orders import place_orders
from typing import List
from agents import (random_agent)
from agents.agent import Agent

# ==============================================================
# Simulation Setup
# ==============================================================

# Initialize the market simulator
sim : market_simulator.Simulator = market_simulator.Simulator(start_time = 0)

# Create some initial orders to seed the market
initial_orders : Orders = Orders(orders=
[
    market_simulator.PendingOrder(
        order_id=1,
        trader_id=0,
        side=market_simulator.OrderSide.BUY,
        quantity=100,
        price=10.00
    ),
    market_simulator.PendingOrder(
        order_id=2,
        trader_id=0,
        side=market_simulator.OrderSide.SELL,
        quantity=100,
        price=15.00
    )
])

# Simulation parameters
run_time : int = 100  # Total simulation time
time_step : int = 1   # Time step for each iteration

# Initialize agents
agents : List[Agent] = [
    random_agent.RandomAgent(name="RandomAgent1", id=1, id_range=range(1000, 8000))
]

# ==============================================================
# Main Simulation Loop
# ==============================================================

def main():

    # Start the simulator with a few initial orders to create market activity
    place_orders(sim=sim, orders=initial_orders)
    sim.submit_pending_orders()

    # Run the simulation loop
    while sim.get_current_time() < run_time:

        # Gather market data for each tick
        market_data : MarketData = MarketData(
            snapshot=sim.get_current_snapshot(),
            level1_data=sim.get_current_level1_data(),
            level2_data=sim.get_current_level2_data()
        )

        # Let each agent update their state and decide on trades
        for agent in agents:

            agent.update(market_data)
            agent.decide_trades()

            orders : Orders = agent.submit_trades()

            # Place orders into the simulator queue
            place_orders(sim, orders)
        
        # Let simulator process orders (randomly chooses which trader to process first)
        sim.submit_pending_orders()
        sim.advance_time(time_step)

    order_logs : List[market_simulator.OrderLog] = sim.get_order_logs()
    trade_logs : List[market_simulator.TradeLog] = sim.get_trade_logs()

    print("Simulation completed.")
    print(f"Total Orders Processed: {len(order_logs)}")
    print(f"Total Trades Executed: {len(trade_logs)}")

    print("Order Logs Sample:", order_logs[:5])
    print("Sample full first trade log:")
    print(trade_logs[0].to_dict() if trade_logs else "No trades executed.")

if __name__ == "__main__":
    main()

File: src/book_implementation/simulation/simulator.cpp
#include "simulator.hpp"

// =============================================
// Simulator Class Implementation
// =============================================S

// Constructor to initialize the simulator with a start time
Simulator::Simulator(Timestamp start_time = 0) {
    simulation_time = start_time;
    
    // Initialize the order book
    order_book.advance_time(simulation_time);
}

// Place a limit order into the simulatorS - only place, do not submit yet
void Simulator::place_limit_order(PendingOrder pending_order) {
    Order order;
    order.order_id = pending_order.order_id;
    order.trader_id = pending_order.trader_id;
    order.price = pending_order.price;
    order.quantity = pending_order.quantity;
    order.side = pending_order.side;
    order.type = OrderType::LIMIT;
    order.timestamp = simulation_time;
    
    pending_orders[pending_order.trader_id] = order;
}

// Place a market order into the simulatorS - only place, do not submit yetS
void Simulator::place_market_order(PendingMarketOrder pending_market_order) {
    Order order;
    order.order_id = pending_market_order.order_id;
    order.trader_id = pending_market_order.trader_id;
    order.price = 0.0; // Price is irrelevant for market orders
    order.quantity = pending_market_order.quantity;
    order.side = pending_market_order.side;
    order.type = OrderType::MARKET;
    order.timestamp = simulation_time;
    
    pending_orders[pending_market_order.trader_id] = order;
}

// Submit all pending orders into the order book
void Simulator::submit_pending_orders() {
    for (const auto& [trader_id, order] : pending_orders) {
        if (order.type == OrderType::LIMIT) {
            order_book.place_limit_order(order);
        } else if (order.type == OrderType::MARKET) {
            order_book.place_market_order(order);
        }
    }
    pending_orders.clear();
}

// Expose current Level 1 market data
Level1Data Simulator::get_current_level1_data() const {
    return order_book.get_level1_data();
}

// Expose current Level 2 market data
Level2Data Simulator::get_current_level2_data() const {
    return order_book.get_level2_data();
}

// Expose current full order book snapshot
OrderBookSnapshot Simulator::get_current_snapshot() const {
    return order_book.get_snapshot(simulation_time);
}

// Advance the simulation time by dt millisecondsS
void Simulator::advance_time(Timestamp dt) {
    simulation_time += dt;
    order_book.advance_time(simulation_time);
}

// Get the current simulation time
Timestamp Simulator::get_current_time() const {
    return simulation_time;
}

// Get all orders for a specific trader
std::vector<Order> Simulator::get_all_trader_orders(TraderID trader_id) const {
    return order_book.get_all_trader_orders(trader_id);
}

// Cancel an existing order by its ID
void Simulator::cancel_order(OrderID order_id) {
    order_book.cancel_order(order_id);
}

// Modify an existing order's price and/or quantity
void Simulator::modify_order(OrderID order_id, Price new_price, Quantity new_quantity) {
    order_book.modify_order(order_id, new_price, new_quantity);
}
